### PR TITLE
vloginov/im2latex ac

### PR DESCRIPTION
**Pull request by morkovka1337**
Add AC support for im2latex model. Implemented custom evaluator and new metric as this model uses metric based on rendering and comparing images. Added annotation converter for im2latex dataset.
Also added two new preprocessing: crop withour resize and resize by scale factor (like fx & fy in cv.resize)
@Ilya-Krylov could you please also review?

### CODE REVIEW

File: models/public/faster_rcnn_inception_v2_coco/faster_rcnn_inception_v2_coco.md [hunk 1]
```diff
@@ -54,7 +54,7 @@ Expected color order: BGR.
    where:
 
     - B - batch size
-    - C - vector of 3 values in format [H,W,S], where H - image height, W - image width, S - image scale factor (usually 1)
+    - C - vector of 3 values in format [H,W,S], where H - image height, W - imahe width, S - image scale factor (usually 1)
```
Comment (vladimir-dudnik) [inline models/public/faster_rcnn_inception_v2_coco/faster_rcnn_inception_v2_coco.md:57]: commit seems not related to this PR

Comment (morkovka1337) [reply models/public/faster_rcnn_inception_v2_coco/faster_rcnn_inception_v2_coco.md]: Rebased to the current develop, could you please recheck this?

**Comment by eaidova** [comment]
Jenkins ok to test

File: tools/accuracy_checker/accuracy_checker/annotation_converters/README.md [hunk 2]
```diff
@@ -329,7 +329,13 @@ Accuracy Checker supports following list of annotation converters and specific f
...
 
 
-## <a name="customizing-dataset-meta"></a>Customizing Dataset Meta
+* `im2latex` - converts im2latex-like datasets to `CharacterRecognitionAnnotation`. [Example of the dataset](http://lstm.seas.harvard.edu/latex/data/)
+  * `images_dir` - path to input images (rendered or scanned formulas)
+  * `formula_file` - path to file containing one formula per line
+  * `split_file` - path to file containing `img_name` and corresponding formula `index` in `formula_file` separated by tab per line
+  * `vocab_file` - file containing vocabulary to cast token class indices into human-readable tokens 
+
+### Customizing dataset meta
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/annotation_converters/README.md:338]:
```suggestion
## <a name="customizing-dataset-meta"></a>Customizing Dataset Meta
```

File: tools/accuracy_checker/accuracy_checker/annotation_converters/im2latex.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
+"""
+Copyright (c) 2019 Intel Corporation
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/annotation_converters/im2latex.py:2]:
```suggestion
Copyright (c) 2018-2020 Intel Corporation
```

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+            progress_reporter.reset(self.dataset.size)
+        self.dataset_meta = self.dataset.metadata
+        for batch_id, (dataset_indices, batch_annotation) in enumerate(self.dataset):
+
+            batch_identifiers = [annotation.identifier for annotation in batch_annotation]
+            batch_input = [self.reader(identifier=identifier) for identifier in batch_identifiers]
+            batch_input = self.preprocessing_executor.process(batch_input, batch_annotation)
+            batch_input, batch_meta = extract_image_representations(batch_input)
+            batch_prediction = self.model.predict(batch_identifiers, batch_input)
+            batch_prediction = [CharacterRecognitionPrediction(label=batch_prediction,
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py:82]: I am right understand, the code support only batch 1?

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+                                                                 identifier=batch_annotation[0].identifier)]
+            self._annotations.append(batch_annotation)
+            self._predictions.append(batch_prediction)
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py:85]: since you concatenate lists, possible extend will be more suitable here

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+        launcher ([type]): [description]
+        suffix ([type]): [description]
+
+    Raises:
+        ValueError: [description]
+
+    Returns:
+        [type]: [description]
+    """
+    launcher_model_mapping = {
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py:206]: not sure that you need factory-like class initialization if you do not plan to support other frameworks for your model

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        raise ValueError('model for framework {} is not supported'.format(framework))
+    return model_class(model_config, launcher, suffix)
+
+
+class SequentialModel:
+    def __init__(self, network_info, launcher, models_args, meta, is_blob=None):
+        recognizer_encoder = network_info.get('recognizer_encoder', {})
+        recognizer_decoder = network_info.get('recognizer_decoder', {})
+        if 'model' not in recognizer_encoder:
+            recognizer_encoder['model'] = models_args[1 if len(models_args) > 1 else 0]
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py:221]: do you have 3 parts in your model or why indexing started with 1?
usually, model args is the list with several paths if you provide in command line
-m <model1> <model2>

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py (hunk 4 above)
Comment (morkovka1337) [reply tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py:82]: Yes, only batch 1 is supported

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py (hunk 5 above)
Comment (morkovka1337) [reply tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py]: Changed to extend

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py (hunk 7 above)
Comment (morkovka1337) [reply tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py]: Fixed this

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py (hunk 6 above)
Comment (morkovka1337) [reply tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py:206]: Actually this function is used to crate both encoder and decoder, but you are right, functionality is excess

File: tools/accuracy_checker/configs/im2latex-0001.yml [hunk 8]
```diff
@@ -0,0 +1,65 @@
+evaluations:
+  - name: im2latex-medium-0002
+    module: custom_evaluators.im2latex_evaluator.Im2latexEvaluator
+    module_config:
+      network_info:
```
Comment (eaidova) [inline tools/accuracy_checker/configs/im2latex-0001.yml:5]: some style errors detected in your config, could you please fix?
```
16:13:16    5:20      error    trailing spaces  (trailing-spaces)
16:13:16    39:1      error    trailing spaces  (trailing-spaces)
16:13:16    59:1      error    trailing spaces  (trailing-spaces)
16:13:16    63:1      error    trailing spaces  (trailing-spaces)
16:13:16    64:1      error    trailing spaces  (trailing-spaces)
16:13:16    65:47     error    no new line character at the end of file  (new-line-at-end-of-file)
```

File: tools/accuracy_checker/configs/im2latex-0001.yml [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    module_config:
+      network_info:  
+        recognizer_encoder:
+          model:   intel/im2latex-medium-scanned-0002/encoder.xml
+          weights: intel/im2latex-medium-scanned-0002/encoder.bin
+
+        recognizer_decoder:
+          model:   intel/im2latex-medium-scanned-0002/decode_step.xml
+          weights: intel/im2latex-medium-scanned-0002/decode_step.bin
+        max_seq_len: '192'
```
Comment (eaidova) [inline tools/accuracy_checker/configs/im2latex-0001.yml:13]: why is it string, not number?

File: tools/accuracy_checker/configs/im2latex-0001.yml [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+              dst_height: 160
+              no_resize: True
+            - type: padding
+              pad_type: right_bottom
+              dst_width: 1400
+              dst_height: 160
+              pad_value: '(255, 255, 255)'
+            - type: normalization
+              std: 255
+              mean: 0
```
Comment (eaidova) [inline tools/accuracy_checker/configs/im2latex-0001.yml:58]: usually, when model published in OMZ as IR, normalization is done during model conversion via MO. if it is your case, please remove.

File: tools/accuracy_checker/accuracy_checker/preprocessor/crop.py [hunk 11]
```diff
@@ -99,22 +104,25 @@ def process_data(data, dst_height, dst_width, central_fraction, use_pillow, is_s
...
-
-        height, width = data.shape[:2]
-        start_height = (height - new_height) // 2
-        start_width = (width - new_width) // 2
-        if is_simple_case:
-            # support GeometricOperationMetadata array for simple case only -- without tiling, pyramids, etc
-            metadata.setdefault('geometric_operations', []).append(GeometricOperationMetadata('crop', {}))
-
-        return data[start_height:start_height + new_height, start_width:start_width + new_width]
+        if not no_resize:
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/preprocessor/crop.py:107]: looks like if no_resize enabled, you do nothing, not only in case, when image small than crop size...
on my view, the check that no resize should be done inside branch  where we check height and width and change new_width or/and new_height to original width and height respectively in case when no_resize enabled

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+from pathlib import Path
+import numpy as np
+
+from accuracy_checker.adapters import create_adapter
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py:19]: please use relative imports

File: tools/accuracy_checker/configs/im2latex-0001.yml (hunk 8 above)
Comment (morkovka1337) [reply tools/accuracy_checker/configs/im2latex-0001.yml]: Fixed

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py (hunk 12 above)
Comment (morkovka1337) [reply tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py]: Done

File: tools/accuracy_checker/accuracy_checker/preprocessor/crop.py (hunk 11 above)
Comment (morkovka1337) [reply tools/accuracy_checker/accuracy_checker/preprocessor/crop.py]: Crop moved into CornerCrop class

File: tools/accuracy_checker/configs/im2latex-0001.yml [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+              pad_value: '(255, 255, 255)'
+
+          metrics:
+            - type: im2latex_match_images_metric
+              max_pixel_column_diff: 0
+
+        - name: im2latex_medium_rendered
+
+          preprocessing:
+            - type: crop
```
Comment (eaidova) [inline tools/accuracy_checker/configs/im2latex-0001.yml:43]: do you not forget update padding here?

File: tools/accuracy_checker/accuracy_checker/preprocessor/resize.py [hunk 14]
```diff
@@ -336,10 +351,13 @@ def parameters(cls):
...
 
     def configure(self):
-        self.dst_height, self.dst_width = get_size_from_config(self.config)
+        allow_none = self.get_value_from_config('aspect_ratio_scale') == 'scale_factor'
+        self.dst_height, self.dst_width = get_size_from_config(self.config, allow_none=allow_none)
         self.resizer = create_resizer(self.config)
         self.scaling_func = ASPECT_RATIO_SCALE.get(self.get_value_from_config('aspect_ratio_scale'))
         self.factor = self.get_value_from_config('factor')
+        if inspect.isclass(self.scaling_func):
+            self.scaling_func = self.scaling_func(self.config, self.parameters)
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/preprocessor/resize.py:360]:
```suggestion
            self.scaling_func = self.scaling_func(self.config, self.parameters())
```

File: tools/accuracy_checker/accuracy_checker/preprocessor/resize.py [hunk 15]
```diff
@@ -92,6 +94,15 @@ def east_keep_aspect_ratio(dst_width, dst_height, image_width, image_height):
 
     return resize_w, resize_h
 
+class ScaleFactor:
+    def __init__(self, config, parameters):
+        self.scale = get_parameter_value_from_config(config, parameters(), 'scale')
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/preprocessor/resize.py:99]:
```suggestion
        self.scale = get_parameter_value_from_config(config, parameters, 'scale')
```

File: tools/accuracy_checker/accuracy_checker/preprocessor/crop.py [hunk 16]
```diff
@@ -415,7 +494,7 @@ def process(self, image, annotation_meta=None):
 class CropOrPad(Preprocessor):
     __provider__ = 'crop_or_pad'
 
-    @classmethod
+    @ classmethod
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/preprocessor/crop.py:497]: why do you add white spaces here???

File: tools/accuracy_checker/configs/im2latex-0001.yml (hunk 13 above)
Comment (morkovka1337) [reply tools/accuracy_checker/configs/im2latex-0001.yml]: Yes, thanks

File: tools/accuracy_checker/accuracy_checker/preprocessor/crop.py (hunk 16 above)
Comment (morkovka1337) [reply tools/accuracy_checker/accuracy_checker/preprocessor/crop.py]: Don't know how it appeared here, deleted it.

**Comment by eaidova** [comment]
@morkovka1337 now, everything looks good, please make pylint happy :)

File: models/intel/im2latex-medium-0001/accuracy-check.yml [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+              dst_width: 1400
+              dst_height: 160
+              pad_value: '(255, 255, 255)'
+
+          metrics:
+            - type: im2latex_match_images_metric
+              max_pixel_column_diff: 0
+
+
+global_definitions: ../dataset_definitions.yml
```
Comment (Anna-Mironova) [inline models/intel/im2latex-medium-0001/accuracy-check.yml:52]: Please remove global_definitions section

Comment (morkovka1337) [reply models/intel/im2latex-medium-0001/accuracy-check.yml]: removed

File: models/intel/im2latex-medium-0001/accuracy-check.yml [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+    module: custom_evaluators.im2latex_evaluator.Im2latexEvaluator
+    module_config:
+      network_info:
+
+        max_seq_len: '192'
+
+      launchers:
+        - framework: dlsdk
+          tags:
+            - FP32
```
Comment (eaidova) [inline models/intel/im2latex-medium-0001/accuracy-check.yml:12]: please remove tag

File: models/intel/im2latex-medium-0001/accuracy-check.yml [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+            - type: im2latex_match_images_metric
+              max_pixel_column_diff: 0
+
+        - name: im2latex_medium_rendered
+
+          preprocessing:
+            - type: corner_crop
+              dst_width: 1400
+              dst_height: 160
+              no_resize: true
```
Comment (eaidova) [inline models/intel/im2latex-medium-0001/accuracy-check.yml:40]: please remove no_resize, here no this option

File: models/intel/im2latex-medium-0001/accuracy-check.yml (hunk 18 above)
Comment (morkovka1337) [reply models/intel/im2latex-medium-0001/accuracy-check.yml]: removed

File: models/intel/im2latex-medium-0001/accuracy-check.yml (hunk 19 above)
Comment (morkovka1337) [reply models/intel/im2latex-medium-0001/accuracy-check.yml]: removed

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py [hunk 20]
```diff
@@ -0,0 +1,302 @@
+"""
+Copyright (c) 2020 Intel Corporation
```
Comment (Anna-Mironova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py:2]: Please align copyright
```suggestion
Copyright (c) 2018-2020 Intel Corporation
```

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+        dec_states_c = enc_res['context']
+        O_t = enc_res['init_0']
+
+        tgt = np.array([[self.sos_index]])
+        logits = []
+        for _ in range(self.max_seq_len):
+
+            dec_res = self.recognizer_decoder.predict(inputs={'row_enc_out': row_enc_out,
+                                                              'dec_st_c': dec_states_c, 'dec_st_h': dec_states_h,
+                                                              'O_t_minus_1': O_t, 'tgt': tgt
```
Comment (Anna-Mironova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py:261]: Decoder model has no 'O_t_minus_1' input
```suggestion
                                                              'output_prev': O_t, 'tgt': tgt
```

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py [hunk 22]
```diff
@@ -0,0 +1,302 @@
...
+                                                              })
+
+            dec_states_h = dec_res['dec_st_h_t']
+            dec_states_c = dec_res['dec_st_c_t']
+            O_t = dec_res['O_t']
```
Comment (Anna-Mironova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/im2latex_evaluator.py:266]: Decoder model has no 'O_t' output
```suggestion
            O_t = dec_res['output']
```

File: tools/accuracy_checker/accuracy_checker/metrics/im2latex_images_match.py [hunk 23]
```diff
@@ -0,0 +1,358 @@
+"""
+Copyright (c) 2020 Intel Corporation
```
Comment (Anna-Mironova) [inline tools/accuracy_checker/accuracy_checker/metrics/im2latex_images_match.py:2]: Please align copyright
```suggestion
Copyright (c) 2018-2020 Intel Corporation
```

File: tools/accuracy_checker/accuracy_checker/metrics/im2latex_images_match.py [hunk 24]
```diff
@@ -0,0 +1,358 @@
...
+    pre_name = output_path.replace('/', '_').replace('.', '_')
+    formula = preprocess_formula(formula)
+    if not os.path.exists(output_path):
+        tex_filename = pre_name + '.tex'
+        log_filename = pre_name + '.log'
+        aux_filename = pre_name + '.aux'
+        with open(tex_filename, "w") as w:
+            w.write(template % formula)
+        run("pdflatex -interaction=nonstopmode {}  >/dev/null".format(tex_filename), TIMEOUT)
+        os.remove(tex_filename)
```
Comment (Anna-Mironova) [inline tools/accuracy_checker/accuracy_checker/metrics/im2latex_images_match.py:135]: please add checks for files existence to avoid attempting to delete non-existent files

```suggestion
        if os.path.exists(tex_filename):
            os.remove(tex_filename)
```

File: tools/accuracy_checker/accuracy_checker/metrics/im2latex_images_match.py [hunk 25]
```diff
@@ -0,0 +1,358 @@
...
+        os.remove(log_filename)
```
Comment (Anna-Mironova) [inline tools/accuracy_checker/accuracy_checker/metrics/im2latex_images_match.py:136]: please add checks for files existence to avoid attempting to delete non-existent files

File: tools/accuracy_checker/accuracy_checker/metrics/im2latex_images_match.py [hunk 26]
```diff
@@ -0,0 +1,358 @@
...
+        os.remove(aux_filename)
```
Comment (Anna-Mironova) [inline tools/accuracy_checker/accuracy_checker/metrics/im2latex_images_match.py:137]: please add checks for files existence to avoid attempting to delete non-existent files

File: tools/accuracy_checker/accuracy_checker/metrics/im2latex_images_match.py [hunk 27]
```diff
@@ -0,0 +1,358 @@
...
+        pdf_filename = tex_filename[:-4] + '.pdf'
+        png_filename = tex_filename[:-4] + '.png'
+        if not os.path.exists(pdf_filename):
+            logging.info('ERROR: %s cannot compile\n', file_idx)
+        else:
+            os.system("convert -density 200 -quality 100 {} {}".format(pdf_filename, png_filename))
+            os.remove(pdf_filename)
```
Comment (Anna-Mironova) [inline tools/accuracy_checker/accuracy_checker/metrics/im2latex_images_match.py:144]: please add checks for files existence to avoid attempting to delete non-existent files

File: models/intel/formula-recognition-medium-scan-0001/accuracy-check.yml [hunk 28]
```diff
@@ -0,0 +1,46 @@
+evaluations:
+  - name: im2latex-medium-0001
```
Comment (Anna-Mironova) [inline models/intel/formula-recognition-medium-scan-0001/accuracy-check.yml:2]:
```suggestion
  - name: formula-recognition-medium-scan-0001
```

File: models/intel/index.md [hunk 29]
```diff
@@ -181,6 +181,7 @@ Deep Learning models for text recognition in various applications.
 | [text-recognition-0012](./text-recognition-0012/description/text-recognition-0012.md)                                                          | 1.485                | 5.568      |
 | [handwritten-score-recognition-0003](./handwritten-score-recognition-0003/description/handwritten-score-recognition-0003.md)                   | 0.792                | 5.555      |
 | [handwritten-japanese-recognition-0001](./handwritten-japanese-recognition-0001/description/handwritten-japanese-recognition-0001.md)          | 117.136              | 15.31      |
+| [im2latex-medium-0001](./im2latex-medium-0001/description/im2latex-medium-0001.md): <br> encoder <br> decoder | <br>16.56<br>1.86 | <br>1.69<br>2.56 |
```
Comment (Anna-Mironova) [inline models/intel/index.md:184]: Please update model's name

Comment (morkovka1337) [reply models/intel/index.md]: updated

File: models/intel/formula-recognition-medium-scan-0001/accuracy-check.yml (hunk 28 above)
Comment (morkovka1337) [reply models/intel/formula-recognition-medium-scan-0001/accuracy-check.yml]: updated

**Review by vladimir-dudnik — CHANGES_REQUESTED** [review]
Why that strange accuracy evaluation approach was chosen? Model output text which you render to image and then compare images. This requires additional applications to be installed (like imagemagick) and may compicate cross platform validation. Could you please change this and compare text with ground true text?

**Review by vladimir-dudnik — APPROVED** [review]
ok, let's go with imagemagick first and see later if there is easier way to do validation

**Comment by vladimir-dudnik** [comment]
@morkovka1337 please resolve conflict

**Comment by eaidova** [comment]
Jenkins please retry a build
